src/magic_folder/status.py

```diff
@@ -368,3 +372,57 @@ def download_finished(self, folder, relpath):
         """
         del self._folders[folder]["downloads"][relpath]
         self._maybe_update_clients()
+
+
+@attr.s
+class _ProxyDescriptor(object):
+    """
+    Descriptor that returns ``partial(self.<original>.<method>, self.<relative>)``.
+    """
+    original = attr.ib(validator=attr.validators.instance_of(unicode))
+    relative = attr.ib(validator=attr.validators.instance_of(unicode))
+    method = attr.ib(validator=attr.validators.instance_of(string_types))
+
+    def __get__(self, oself, type=None):
+        if oself is None:
+            raise NotImplementedError()
+        original = getattr(oself, self.original)
+        return partial(
+            getattr(original, self.method),
+            getattr(oself, self.relative),
+        )
+
+def relativeProxyFor(iface, original, relative):
+    """
+    Class decorator that partially applies methods of an interface.
+
+    For each method of the given interface, that takes as first argument the
+    given relative argument, this adds a corresponding proxy method to the
+    decorated class, that gets that argument from the decorated instance.
+
+    :param Interface iface: The interface to generate proxy methods for.
+    :param unicode original: The attribute on the decorated class that
+        contains the implementation of the interface.
+    :param unicode relative: The attribute on the decorated class to
+        pass as the first argument to methods of the interface, when
+        the name of the first argument matches.
+    """
+    def decorator(cls):
+        for name, method in iface.namesAndDescriptions():
+            if not isinstance(method, Method):
+                continue
+            if method.positional[0] == relative:
+                setattr(cls, name, _ProxyDescriptor(original, relative, name))
+        return cls
+    return decorator
+
+
+@relativeProxyFor(IStatus, "_status", "folder")
```

Comment:
I see where you're going with this .. but now we're tying _all_ (current and future) "status" APIs to require a "folder" as the first argument.
It also makes this code somewhat hard to read? (e.g. instead of having an `IStatus` interface to look at, you have to trace back here and figure out that it's an `IStatus` under the hood with magic first arg before then finding `IStatus` to see what it actually does).

I guess overall I'm "-0" on making this so generic for a single use-case (i.e. just to avoid accessing the right folder-name for the half-dozen call-sites that require those).

Reply:
> but now we're tying all (current and future) "status" APIs to require a "folder" as the first argument.

I very deliberately designed this to *not* constrain us that way. This *only* proxies method that take `folder` as the first argument, and ignores methods that don't have that as a first argument (and non-methods). While that does limit what is available via `FolderStatus`, `IStatus` can have other methods that are accessible directly.

The current implementation keeps the underlying `IStatus` private, so we can't access those methods from a `FolderStatus`, but my feeling is we shouldn't need to (and we can always revisit that).

My immediate motivation for this was wanting to pass a `FolderStatus` to a function that doesn't currently have access to the name, without also needing to pass the name in (`magic_folder.scanner.find_updated_files`). I also noticed when the status code initially landed that you changed at least one object around that *didn't* have the name, just to be able to report status. I considered suggesting something like this (at least the interface, not necessarily the implementation).

Reply:
>  I also noticed when the status code initially landed that you changed at least one object around that didn't have the name, just to be able to report status.

I feel like "approximately everything" is fair game to have access to "the config" so don't see that as a problem.

Reply:
> This only proxies method that take folder as the first argument, and ignores methods that don't have that as a first argument (and non-methods). 

Hmm, okay I see that now ..  that's good.